File: Lab1/search_pthread.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <time.h>

#define N 100000000
#define TARGET 42
#ifndef NUM_THREADS
#define NUM_THREADS 8
#endif

int *arr;
int found_index = -1;
pthread_mutex_t lock = PTHREAD_MUTEX_INITIALIZER;

typedef struct {
    long start;
    long end;
} ThreadArgs;

void* search_chunk(void* arg) {
    ThreadArgs *args = (ThreadArgs*) arg;
    for (long i = args->start; i < args->end; i++) {
        // Stop if found
        pthread_mutex_lock(&lock);
        int stop = (found_index != -1);
        pthread_mutex_unlock(&lock);
        if (stop) break;

        if (arr[i] == TARGET) {
            pthread_mutex_lock(&lock);
            if (found_index == -1) found_index = i; // first occurrence
            pthread_mutex_unlock(&lock);
            break;
        }
    }
    return NULL;
}

int main() {
    // Allocate and fill array
    arr = malloc(N * sizeof(int));
    if (!arr) { perror("malloc"); return 1; }
    srand(time(NULL));
    for (long i = 0; i < N; i++) arr[i] = rand() % 100;

    pthread_t threads[NUM_THREADS];
    ThreadArgs args[NUM_THREADS];
    long chunk_size = N / NUM_THREADS;

    // Start timing
    clock_t start = clock();

    // Create threads
    for (int t = 0; t < NUM_THREADS; t++) {
        args[t].start = t * chunk_size;
        args[t].end = (t == NUM_THREADS-1) ? N : (t+1) * chunk_size;
        pthread_create(&threads[t], NULL, search_chunk, &args[t]);
    }

    // Join threads
    for (int t = 0; t < NUM_THREADS; t++) pthread_join(threads[t], NULL);

    clock_t end = clock();
    double elapsed = (double)(end - start) / CLOCKS_PER_SEC;

    printf("Found at index = %d\n", found_index);
    printf("Time with %d threads = %.6f seconds\n", NUM_THREADS, elapsed);

    free(arr);
    pthread_mutex_destroy(&lock);
    return 0;
}

File: Lab1/search_pthread.py
import random
import multiprocessing as mp
import time

N = 100_000_000
TARGET = 42
NUM_PROCESSES = 8

# Create array in main process
arr = [random.randint(0, 99) for _ in range(N)]
found_index = mp.Value('i', -1)

def search_chunk(start, end):
    for i in range(start, end):
        if found_index.value != -1:
            break
        if arr[i] == TARGET:
            with found_index.get_lock():
                if found_index.value == -1:
                    found_index.value = i
            break

if __name__ == "__main__":
    chunk_size = N // NUM_PROCESSES
    processes = []

    start_time = time.time()

    for t in range(NUM_PROCESSES):
        s = t * chunk_size
        e = N if t == NUM_PROCESSES-1 else (t+1) * chunk_size
        p = mp.Process(target=search_chunk, args=(s, e))
        processes.append(p)
        p.start()

    for p in processes:
        p.join()

    end_time = time.time()

    print(f"Found at index = {found_index.value}")
    print(f"Time with {NUM_PROCESSES} processes = {end_time - start_time:.2f} seconds")
